NEXUS-4735 as initial reason, but many (but less than previously) IT fixes

cstamas opened the pull request:
The inital problem (log file kept open on windows making it impossible to delete) led me to believe there was a component in IT plexus container that kept the file open.

Plexus is now bounced again per-IT-class, just as before.

Also, how security was managed in ITs was next to a disaster, and it turned out that many ITs did "interleave" (was relying on side effect!).

After fixing security, next was to remove the "feature" of ITs that it was bringin' up almost all of Nexus internals within IT (to "validate" configuration [why?], to mangle security configuration [why?] and many more). So, alternative implementations are added that does NOT bring up all of those components, except in few (2 i believe) ITs where they are a must to decipher some of the SMTP related passwords that would come in scrambled if using "low level" model loaded only.

Finally, it turned out that a few ITs (within same group) was responsible for more than 20minutes of runtime, and they were even not doing their job: they were silently timeouting (auto block related ITs).

cstamas committed:
NEXUS-4735: fix

This is not fixing the culprit, but I assume we will not see that problem
again (happened once on CI win slave).

cstamas committed:
NEXUS-4735: fixes and more

The initial problem seems fixed, but I went over improving ITs, and it turned out, that
many of ITs had side effects it was misusing. Also, one complete set of ITs was never working (!)
but they were simply "passing" as it was timing out after 10 minutes or run!
Also, plexus was bounced per-IT, while there was no need for it at all (all the "IT plexus" is used
is to look up some helper clients and so).

cstamas committed:
NEXUS-4735: muting one NEXUS-4635 test as it does not make sense.

cstamas committed:
Typo fix

cstamas committed:
ThresdPoolManager leak fix.

cstamas committed:
Getting rid of unneded warning.

cstamas committed:
NEXUS-4735: undoing the "do not bounce" Plexus.

Now again, one Plexus per IT class is started and then stopped.
Also, a slight change creeped in the ProgressListener, undoing it.

cstamas committed:
NEXUS-4735: the only IT that was using the "a la plexus config" feature.

is changed to not use it anymore.

bdemers commented on nexus/nexus-configuration/src/main/java/org/sonatype/nexus/configuration/application/upgrade/Upgrade146to1100.java:
```diff
@@ -102,6 +104,18 @@ public CRepository upgradeCRepository( org.sonatype.nexus.configuration.model.v1
 
         org.sonatype.nexus.configuration.model.Configuration newc = versionConverter.upgradeConfiguration( oldc );
+        
+        // this should go into much "older" upgrader, this was a mistake!
```
why not put it in an older upgrader ?

cstamas replied:
because older one will eventually never execute (think X-1 > X upgrade). Ultimately, an "upgrader" component that went out once with a release X should _never_ be modified in any of the Y (X < Y) release. Hence, it's always the _latest_ upgrader that is changed (even if we spot that some older upgrader is bad _after_ we released it(.

bdemers replied:
can you add a note in the comment, about this.  So no one will read the comment and think it should be moved in the future.

bdemers commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/repository/threads/ThreadPoolManager.java:
```diff
@@ -29,4 +29,6 @@
 
     void removePool( Repository repository );
+    
+    // void shutdown();
```
dead code

cstamas replied:
true, i was toying with idea to drop the Plexus specific lifecycle (Disposable et al). But in last moment, I gave up, did not want to disturb it in this changeset.

bdemers commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/repository/threads/ThreadPoolManagerEventInspector.java:
```diff
@@ -37,12 +42,18 @@
     {
-        return evt instanceof RepositoryRegistryRepositoryEvent;
+        return evt != null && evt instanceof RepositoryRegistryRepositoryEvent;
+        // return evt != null && ( evt instanceof RepositoryRegistryRepositoryEvent || evt instanceof NexusStoppedEvent );
```
dead code

cstamas replied:
Left here for reference (when we leave plexus apis completely).

bdemers commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/repository/threads/ThreadPoolManagerEventInspector.java:
```diff
@@ -52,5 +63,9 @@ else if ( evt instanceof RepositoryRegistryEventRemove )
+        // else if ( evt instanceof NexusStoppedEvent )
+        // {
+        // poolManager.shutdown();
+        // }
```
dead code

bdemers commented on nexus/nexus-test-harness/nexus-test-harness-launcher/src/main/java/org/sonatype/nexus/integrationtests/AbstractNexusIntegrationTest.java:
```diff
@@ -350,25 +341,26 @@ public void oncePerClassSetUp()
-                // start nexus
-                startNexus();
+                // set security enabled/disabled as expected by current IT
+                TestContainer.getInstance().invokeAsAdministrator( new Callable<Object>()
```
This is a little verbose.
But it is MUCH better then having the test deal with which user to use for requests!

bdemers commented on nexus/nexus-test-harness/nexus-test-harness-launcher/src/main/java/org/sonatype/nexus/integrationtests/AbstractNexusIntegrationTest.java:
```diff
@@ -930,24 +920,26 @@ private static void takeSnapshot()
+     * @return
+     * @deprecated Use {@link TestContainer#getBasedir()} instead.
+     */
+    @Deprecated
```
why deprecate this?  this seems useful rather then calling TestContainer.getBasedir()
Why not just leave this method in place?

cstamas replied:
I like to have _one source of truth_. Having two public methods that returns basedir will confuse devs in future.

Having this non-deprecated will put us in that situation: two methods seemingly doing the same....

bdemers replied:
I agree with the one source of true.  Its just that for the longest time people have been used to this method in various test support classes (PlexusTestCase.getBasedir() for example)

if you really feel strong about it, leave it deprecated

bdemers commented on nexus/nexus-test-harness/nexus-test-harness-launcher/src/main/java/org/sonatype/nexus/integrationtests/AbstractNexusIntegrationTest.java:
```diff
@@ -930,24 +920,26 @@ private static void takeSnapshot()
+        return TestContainer.getBasedir();
     }
 
+    @Deprecated
```
add @deprecated javadoc

bdemers commented on nexus/nexus-test-harness/nexus-test-harness-launcher/src/main/java/org/sonatype/nexus/test/utils/NexusConfigUtil.java:
```diff
@@ -33,14 +36,28 @@
+ * 
+ * @author cstamas
+ */
 public class NexusConfigUtil
```
Good comments

bdemers commented:
Other then some minor stuff +1

adreghiciu commented:
+1

cstamas committed:
NEXUS-4735: fix

This is the real culprit of the issue originally reported. Logback logfiles are released when logback instance
is GCed. This is obviously not seen in "normal" ITs since they log in separate place, but in this case, immediately
after a shutdown, next IT run failed since -- probably -- the GC of previous one did not ran/finish/whatever,
but we wanted immediately to start another IT that begins with deleting the folder. Again, why we never see this
in case of other ITs is that every IT logs to it's own "specific" place, not affected by "clean workdir" step of
preparations for IT run.

cstamas committed:
Now that core ITs are cleaned up, returning the a'la Plexus configability,
that is required by some Pro ITs.

cstamas commented:
CI +1:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/151/

cstamas committed:
As some ITs expect, moving Plexus creation to happen before oncePerClass().

velo commented on nexus/nexus-configuration/src/main/java/org/sonatype/nexus/configuration/source/FileConfigurationSource.java:
```diff
@@ -195,6 +196,8 @@ public Configuration loadConfiguration()
         ValidationResponse vResponse =
             getConfigurationValidator().validateModel( new ValidationRequest( getConfiguration() ) );
+        
+        dumpValidationErrors( vResponse );
```
AWESOME!!!

velo replied:
ow, I though there was no dump before, looks like there was =D

velo commented:
+1
